planner/core/rule_inject_extra_projection.go

```diff
@@ -62,10 +63,42 @@ func (pe *projInjector) inject(plan PhysicalPlan) PhysicalPlan {
 		plan = InjectProjBelowSort(p, p.ByItems)
 	case *NominalSort:
 		plan = TurnNominalSortIntoProj(p, p.OnlyColumn, p.ByItems)
+	case *PhysicalUnionAll:
+		plan = injectProjBelowUnion(p)
 	}
 	return plan
 }
 
+func injectProjBelowUnion(un *PhysicalUnionAll) *PhysicalUnionAll {
```

Comment:
Is it necessary? In the logical plan building stage, field cast projection has already been appended.

Reply:
It is! The check in logical plan is looser, such as lack of "NotNull" flag check.